### PR TITLE
[PIPE-316] Recipient Profile Databricks

### PR DESCRIPTION
**Pull request by emily-brents**
**Description:**
Adapts the `restock_recipient_profile` sql script to wirk in databricks.

**Requirements for PR merge:**

1. [X] Unit & integration tests updated
2. [X] API documentation updated
3. [ ] Necessary PR reviewers:
    - [ ] Backend
    - [ ] Frontend
4. [X] Matview impact assessment completed
5. [X] Frontend impact assessment completed
6. [ ] Data validation completed
7. [ ] Appropriate Operations ticket(s) created
8. [X] Jira Ticket [PIPE-316](https://federal-spending-transparency.atlassian.net/browse/PIPE-316):
    - [X] Link to this Pull-Request
    - [ ] Performance evaluation of affected (API | Script | Download)
    - [ ] Before / After data comparison

**Area for explaining above N/A when needed:**
```
```

### CODE REVIEW

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (sethstoudenmier) [inline usaspending_api/recipient/delta_models/recipient_profile.py:37]: Rest of the comments are Python comments so I would change this to match.

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (sethstoudenmier) [inline usaspending_api/recipient/delta_models/recipient_profile.py:28]: Any reason to keep this around?

File: usaspending_api/etl/tests/integration/test_load_to_from_delta.py
Comment (sethstoudenmier) [inline usaspending_api/etl/tests/integration/test_load_to_from_delta.py:771]: Why is this commented out? Is it needed?

File: usaspending_api/etl/management/commands/load_query_to_delta.py
Comment (sethstoudenmier) [inline usaspending_api/etl/management/commands/load_query_to_delta.py:10]: Not worth blocking over, but can you follow pattern of alphabetical ordering?

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (sethstoudenmier) [inline usaspending_api/recipient/delta_models/recipient_profile.py:435]: Don't think you need to drop the view since it will only exist for the duration of the session.

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (sethstoudenmier) [inline usaspending_api/recipient/delta_models/recipient_profile.py:99]: Does a location need to be specified for the temporary table? Also probably good to specify which DB the table will be created in.

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (sethstoudenmier) [reply usaspending_api/recipient/delta_models/recipient_profile.py:435]: Also, might need to add the DB you are dropping the table from depending on the response to question regarding where the table is created.

**Review by sethstoudenmier — CHANGES_REQUESTED** [review]
Most comments around some suggested cleanup. There are some additional comments around the creation and deletion of the temporary table and view. Didn't look too much at the SQL that was copied over quite yet, but plan to once Travis is happy.

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (emily-brents) [reply usaspending_api/recipient/delta_models/recipient_profile.py]: Nope, removed

File: usaspending_api/etl/tests/integration/test_load_to_from_delta.py
Comment (emily-brents) [reply usaspending_api/etl/tests/integration/test_load_to_from_delta.py]: uncommented

File: usaspending_api/etl/management/commands/load_query_to_delta.py
Comment (emily-brents) [reply usaspending_api/etl/management/commands/load_query_to_delta.py:10]: Re-ordered import statements

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (sethstoudenmier) [inline usaspending_api/recipient/delta_models/recipient_profile.py:96]: My understanding is that `monotonically_increasing_id()` is guaranteed unique for a single query / table (assuming it falls within the 8 billion constraint for partitions), but that the value is not guaranteed unique between tables / queries.

Since you have `monotonically_increasing_id() AS id` that means you could have multiple of the same `id` field in this query with UNION statements. 

For example, 
```
SELECT monotonically_increasing_id()
UNION ALL 
SELECT monotonically_increasing_id();
```
yields 
```
0
0
```

File: usaspending_api/tests/conftest_spark.py
Comment (sethstoudenmier) [inline usaspending_api/tests/conftest_spark.py:207]: Why remove the CASCADE? Should be able to keep this place.

**Review by sethstoudenmier — COMMENTED** [review]
Biggest question on the PR is about `monotonically_increasing_id()` and the effects it might have after multiple MERGE INTO of rpt.recipient_profile.

File: usaspending_api/etl/management/commands/load_query_to_delta.py
Comment (collinwr) [inline usaspending_api/etl/management/commands/load_query_to_delta.py:61]: Could you make this plural to give a hint to the reader that this is a list of strings?

File: usaspending_api/tests/conftest_spark.py
Comment (emily-brents) [reply usaspending_api/tests/conftest_spark.py]: That was a test i did that shouldn't have been pushed, replaced

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (emily-brents) [reply usaspending_api/recipient/delta_models/recipient_profile.py:96]: There isn't a way to duplicate the auto-incrementing ids from postgres, so for the time being I am just mocking the IDs in delta, and when we move the data back into postgres we should be able to not move the id field and let the postgres table automatically fill in the id field. We don't use the id field for anything as far as i'm aware, so it should be fine for it to be dummy data.

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (collinwr) [inline usaspending_api/recipient/delta_models/recipient_profile.py:348]: What's the purpose of populating a temporary table, `temporary_restock_recipient_profile`, then performing a series of merges to make `recipient_profile` match? 
I believe this was done in Postgres to prevent a live table from being updated by a series of steps (1-9) that could fail, but because this DBR ETL is updating a table that will later be nuke/paved, would it make sense to clear `recipient_profile` at the start of the script and update directly?

Comment (emily-brents) [reply usaspending_api/recipient/delta_models/recipient_profile.py:348]: yeah that is fair, i'll see what i can do

File: usaspending_api/etl/management/commands/load_query_to_delta.py
Comment (emily-brents) [reply usaspending_api/etl/management/commands/load_query_to_delta.py]: sure

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (kwhickey) [reply usaspending_api/recipient/delta_models/recipient_profile.py:30]: Seeing the whole review, there's not one major INSERT, but a lot of MERGEs. I think you could still save this delete step and the first INSERT be an `INSERT OVERWRITE` which implicitly performs a TRUNCATE. But not sure if it will make much difference in speed and may make the code less clear.

**Review by kwhickey — CHANGES_REQUESTED** [review]
Overall looking solid in code, SQL, and tests.  A few areas to address or talk through:
1. labeling each sql step run
2. what do we choose for the partition_column and do we need it?
3. Comment about not having to duplicate all tests for new tables
4. Lots to ponder about doing all overlaying "MERGE" transformation in memory vs. causing delta writes of parquet files

File: usaspending_api/etl/tests/integration/test_load_to_from_delta.py
Comment (kwhickey) [inline usaspending_api/etl/tests/integration/test_load_to_from_delta.py:503]: So if this is a base function used in the setup of many of the other unit tests ... it's going to "create + load" the Delta table for `recipient_profile` using the resident `data_store_api.rpt.recipient_profile` _postgres_ data, yes? 

Really with your changes, `recipient_profile` delta table shouldn't be just a copy from postgres. But "loading" it should be done via the `load_query_to_delta`, which builds it up from other base delta tables. 

Maybe this is all deferrable til when `transaction_search` and `award_search` spark sql start referencing `rpt.recipient_profile` (and `rpt.recipient_lookup`) delta tables instead of the `raw.*` ones? If so, then you need to make sure that at that time we effectively blow away `raw.recipient_profile` and refactor these unit tests to build-up delta tables in dependency order (according to the vertical slices in our diagram):
1. copy base delta table data over from postgres
2. create "step 2" delta tables (e.g. `rpt.recipient_lookup`)
3. create "step 3" delta tables (e.g. `rpt.recipient_profile`)
4. create "step 4" delta tables (e.g. `rpt.*_search`)

File: usaspending_api/etl/management/commands/load_query_to_delta.py
Comment (kwhickey) [inline usaspending_api/etl/management/commands/load_query_to_delta.py:75]: Please add the new column underneath: 
`"is_partition_column_unique": False,`

**Review by kwhickey — CHANGES_REQUESTED** [review]
Almost. SQL looks good. There's 1 changes and 2 questions:
1. when does the SQL start using `rpt.recipient_lookup` (see comment from previous review)
2. note to future self about refactoring test delta table buildup

Changes
1. adding the is_partition_column_unique to the spec

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (emily-brents) [reply usaspending_api/recipient/delta_models/recipient_profile.py:443]: removed the table, need to drop view to prevent errors with the tests.

File: usaspending_api/etl/management/commands/load_table_to_delta.py
Comment (emily-brents) [reply usaspending_api/etl/management/commands/load_table_to_delta.py:102]: added testing table

File: usaspending_api/etl/management/commands/load_query_to_delta.py
Comment (emily-brents) [reply usaspending_api/etl/management/commands/load_query_to_delta.py]: Added logging statement

File: usaspending_api/etl/tests/integration/test_load_to_from_delta.py
Comment (emily-brents) [reply usaspending_api/etl/tests/integration/test_load_to_from_delta.py]: removed alt-name and alt-db tests

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (emily-brents) [reply usaspending_api/recipient/delta_models/recipient_profile.py]: refactored

File: usaspending_api/recipient/delta_models/recipient_profile.py
Comment (emily-brents) [reply usaspending_api/recipient/delta_models/recipient_profile.py:93]: Changing this to `rpt.recipient_lookup` depends on getting https://github.com/fedspendingtransparency/usaspending-api/pull/3541 merged in, otherwise changing this will cause errors.

File: usaspending_api/etl/tests/integration/test_load_to_from_delta.py
Comment (emily-brents) [reply usaspending_api/etl/tests/integration/test_load_to_from_delta.py:503]: Initially I had the changes to `award_search` and `transaction_search` as part of this PR, but was told to remove them, I can re-do those changes and make the tests use the new `rpt.recipient_profile, I'm just getting mixed messages.

File: usaspending_api/etl/management/commands/load_query_to_delta.py
Comment (emily-brents) [reply usaspending_api/etl/management/commands/load_query_to_delta.py:75]: added

File: usaspending_api/etl/tests/integration/test_load_to_from_delta.py
Comment (kwhickey) [reply usaspending_api/etl/tests/integration/test_load_to_from_delta.py:503]: I think it's just a future thing (next-sprint). So you can hold for now, but it'll be a confusing refactor in the future.

**Review by kwhickey — DISMISSED** [review]
Approving with latest

File: usaspending_api/etl/management/commands/load_table_to_delta.py
Comment (sethstoudenmier) [inline usaspending_api/etl/management/commands/load_table_to_delta.py:113]: Shouldn't this still be `True`?

Comment (emily-brents) [reply usaspending_api/etl/management/commands/load_table_to_delta.py:113]: No, recipient_profile has `recipient_hashes` for potentially all levels of any recipient hash - i.e., `recipient_hash-C` `recipient_hash-R`, `recipient_hash-P`

Comment (sethstoudenmier) [reply usaspending_api/etl/management/commands/load_table_to_delta.py:113]: The partition_column in this TABLE_SPEC is the `id` though not the hash

Comment (emily-brents) [reply usaspending_api/etl/management/commands/load_table_to_delta.py:113]: oh fair, fixed

**Review by sethstoudenmier — APPROVED** [review]

**Review by collinwr — APPROVED** [review]
My feedback was addressed. Approving

**Pull request merged**
